oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java

```diff
@@ -1252,6 +1282,37 @@ public void removeGarbage(final VersionGCStats stats) {
                 delayOnModifications(timer.stop().elapsed(MILLISECONDS), cancel);
             }
         }
+
+        private boolean verify(NodeDocument oldDoc, NodeDocument newDoc, UpdateOp update) {
```

Comment:
This comparison is doing more than required. In case the oldDoc is null, we should ignore it and move forward cause that is already taken care of by Mongo.

Mongo only updates if it finds the document else, it ignores the update operation.

Reply:
the comparison is at the `DocumentNodeState` level though, not `NodeDocument`. oldDoc and newDoc are both not null here

Reply:
What I meant is that https://github.com/apache/jackrabbit-oak/pull/1262/files#diff-e47c4905644873320ba8f5ed72096b1c6b2ad3fdec6d51241b27a413c371562aR1297-R1303

In the above code, if `oldNs` is null then we should ignore and return true.

Reply:
can you elaborate why?

Reply:
I thought, _because null would mean that the document has been deleted and in that case, it doesn't make sense to compare them cause it has already been gone._

But now I think you are right. Deleted documents can be resurrected or that commit can be reverted and we do need to compare them.

Reply:
Yes, I think we should compare for all variations there. The _node_ can be deleted in the old or new state - but it has to be the same, otherwise `removeGarbage` cleaned up some wrong garbage